changefeedccl: webhook_sink_config option to configure batching and retries

spiffyy99 opened the pull request:
Previously, retry behavior was preset and batching was not supported. This
change adds configuration JSON to change retry behavior and batch sink
messages by count, bytes, and timeouts.

Resolves #67804

Release note (enterprise change): webhook_sink_config JSON option to
configure retry and batching behavior for webhook sink changefeed messages

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/68633)
<!-- Reviewable:end -->

stevendanna reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68633)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @spiffyyeng)

___
*[pkg/ccl/changefeedccl/sink_webhook.go, line 57 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68633#-Mgk6KEJDfMFIBJNgOU4:-Mgk6KEJDfMFIBJNgOU5:b5dvd7h) ([raw file](https://github.com/cockroachdb/cockroach/blob/770aace7865c5ded1d3dd13d45b7a058ce39f2c1/pkg/ccl/changefeedccl/sink_webhook.go#L57)):*
> ```Go
> 
> type webhookSinkPayload struct {
> 	Payload []json.RawMessage `json:"payload"`
> ```

It might be nice to add a count of the messages contained in Payload.
___
*[pkg/ccl/changefeedccl/sink_webhook.go, line 61 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68633#-Mgk60JS-42lTmFT6p-U:-Mgk60JS-42lTmFT6p-V:b6mdwry) ([raw file](https://github.com/cockroachdb/cockroach/blob/770aace7865c5ded1d3dd13d45b7a058ce39f2c1/pkg/ccl/changefeedccl/sink_webhook.go#L61)):*
> ```Go
> 
> func encodePayloadWebhook(values [][]byte) ([]byte, error) {
> 	var payload []json.RawMessage
> ```

nit: since we know the len of values, perhaps use make here.
___
*[pkg/ccl/changefeedccl/sink_webhook.go, line 77 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68633#-Mgk6UXF6SEuIUVKdObE:-Mgk6UXF6SEuIUVKdObF:b-sirksg) ([raw file](https://github.com/cockroachdb/cockroach/blob/770aace7865c5ded1d3dd13d45b7a058ce39f2c1/pkg/ccl/changefeedccl/sink_webhook.go#L77)):*
> ```Go
> type batchTracker struct {
> 	buffer                  [][]byte
> 	bufferBytes, bufferSize int
> ```

Maybe `bufferMsgCount` rather than `bufferSize`?
___
*[pkg/ccl/changefeedccl/sink_webhook.go, line 92 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68633#-MgjqV9c5CckovB5f9Z0:-MgjqV9c5CckovB5f9Z1:bexz3bo) ([raw file](https://github.com/cockroachdb/cockroach/blob/770aace7865c5ded1d3dd13d45b7a058ce39f2c1/pkg/ccl/changefeedccl/sink_webhook.go#L92)):*
> ```Go
> 	b.bufferSize = 0
> 	b.bufferBytes = 0
> 	b.bufferInitTime = timeutil.Now()
> ```

We may want to set this on the first write into the buffer so that a gap between two sets of messages doesn't mean that the first message to get written immediately gets sent.
___
*[pkg/ccl/changefeedccl/sink_webhook.go, line 122 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68633#-Mgk9zMY9F4i21lz4l9I:-Mgk9zMY9F4i21lz4l9J:b8qss8b) ([raw file](https://github.com/cockroachdb/cockroach/blob/770aace7865c5ded1d3dd13d45b7a058ce39f2c1/pkg/ccl/changefeedccl/sink_webhook.go#L122)):*
> ```Go
> 	opts map[string]string,
> ) (batchCfg batchConfig, retryCfg retry.Options, err error) {
> 	defaultRetryCfg := defaultRetryConfig()
> ```

I would maybe rename this variable `retryCfg` since we mutate it later in this function.
___
*[pkg/ccl/changefeedccl/sink_webhook.go, line 335 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68633#-Mgk8FLY-E0pCQptBZ45:-Mgk8FLZ9xL72awFpBzE:b-96qy4y) ([raw file](https://github.com/cockroachdb/cockroach/blob/770aace7865c5ded1d3dd13d45b7a058ce39f2c1/pkg/ccl/changefeedccl/sink_webhook.go#L335)):*
> ```Go
> 	// 1. The default batch behavior is in place (send every time)
> 	// 2. Any one of the conditions are met UNLESS the condition is zero which means never batch
> 	return s.batchDefault ||
> ```

Do we need this batchDefault boolean?  This will only be true if the user didn't provide config.  But, if the user didn't provide configuration, then Messages, Bytes, and Frequency should all be zero-valued so all of the conditions below will be false.


<!-- Sent from Reviewable.io -->

stevendanna reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68633)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @spiffyyeng)

___
*[pkg/ccl/changefeedccl/sink_webhook.go, line 335 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68633#-Mgk8FLY-E0pCQptBZ45:-MgkZG8f1Q_EcMZ20xTV:b-l4l3zf) ([raw file](https://github.com/cockroachdb/cockroach/blob/770aace7865c5ded1d3dd13d45b7a058ce39f2c1/pkg/ccl/changefeedccl/sink_webhook.go#L335)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

Do we need this batchDefault boolean?  This will only be true if the user didn't provide config.  But, if the user didn't provide configuration, then Messages, Bytes, and Frequency should all be zero-valued so all of the conditions below will be false.
</blockquote></details>

Backfill from conversation: Apologies for leaving a little bit of a non-sensical comment here.

What I meant to say: I think we might want to pursue a configuration scheme where `s.batchDefault` can be replaced with `s.batchCfg == batchConfig{}`.  Since the all-zero configuration would otherwise lead to a batcher that never sends anything per the below condition, it is then open for being used to mean "never batch, send immediately."

As came up in our discussion, we likely also want some checks against configurations that could lead to batches never being sent (such as Bytes or Messages being set but no Frequency being set).


<!-- Sent from Reviewable.io -->

spiffyyeng commented:
A few things to note about batching behavior (we should elaborate on this during docs writing):
- Externally, we refer to batching as "Flushing" in line with Sarama. We refer to it internally as batching to be distinct from the flushing behavior expected in the Sink interface. 
- There are a few edge cases to take note of when it comes to batching config:
```
  "Flush": {
    "Messages":  ...,
    "Bytes":     ...,
    "Frequency": ...,
  },
```
    - If no batching config is provided, all the values will be set to zero. this is equivalent to batching every time.
    - if at least one value is nonzero, any zero values will be treated as "unlimited/infinite", and will never trigger a batch to be sent.
    - if the frequency is not set, and either the messages or bytes are sent, we issue a warning (since the batch may never send)
- Right now, batches are sent per worker. We should communicate this as a per-key batching guarantee, or that messages with the same key will be considered for the same batch.

spiffyy99 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68633)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @stevendanna)

___
*[pkg/ccl/changefeedccl/sink_webhook.go, line 57 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68633#-Mgk6KEJDfMFIBJNgOU4:-Mgl4pT6D9Diy_F9AbHW:b-ao9t2k) ([raw file](https://github.com/cockroachdb/cockroach/blob/770aace7865c5ded1d3dd13d45b7a058ce39f2c1/pkg/ccl/changefeedccl/sink_webhook.go#L57)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

It might be nice to add a count of the messages contained in Payload.
</blockquote></details>

Done. (called size)
___
*[pkg/ccl/changefeedccl/sink_webhook.go, line 77 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68633#-Mgk6UXF6SEuIUVKdObE:-Mgl4rc17WflEpdqkC66:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/770aace7865c5ded1d3dd13d45b7a058ce39f2c1/pkg/ccl/changefeedccl/sink_webhook.go#L77)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

Maybe `bufferMsgCount` rather than `bufferSize`?
</blockquote></details>

Done.
___
*[pkg/ccl/changefeedccl/sink_webhook.go, line 92 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68633#-MgjqV9c5CckovB5f9Z0:-Mgl4t42APejp1YadtBX:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/770aace7865c5ded1d3dd13d45b7a058ce39f2c1/pkg/ccl/changefeedccl/sink_webhook.go#L92)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

We may want to set this on the first write into the buffer so that a gap between two sets of messages doesn't mean that the first message to get written immediately gets sent.
</blockquote></details>

Done.
___
*[pkg/ccl/changefeedccl/sink_webhook.go, line 122 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68633#-Mgk9zMY9F4i21lz4l9I:-Mgl4tmuEj-d6aFOktBG:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/770aace7865c5ded1d3dd13d45b7a058ce39f2c1/pkg/ccl/changefeedccl/sink_webhook.go#L122)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

I would maybe rename this variable `retryCfg` since we mutate it later in this function.
</blockquote></details>

Done.
___
*[pkg/ccl/changefeedccl/sink_webhook.go, line 335 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68633#-Mgk8FLY-E0pCQptBZ45:-Mgl4ufLBn2BkWf0yAyE:b-7p6d1) ([raw file](https://github.com/cockroachdb/cockroach/blob/770aace7865c5ded1d3dd13d45b7a058ce39f2c1/pkg/ccl/changefeedccl/sink_webhook.go#L335)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

Backfill from conversation: Apologies for leaving a little bit of a non-sensical comment here.

What I meant to say: I think we might want to pursue a configuration scheme where `s.batchDefault` can be replaced with `s.batchCfg == batchConfig{}`.  Since the all-zero configuration would otherwise lead to a batcher that never sends anything per the below condition, it is then open for being used to mean "never batch, send immediately."

As came up in our discussion, we likely also want some checks against configurations that could lead to batches never being sent (such as Bytes or Messages being set but no Frequency being set).
</blockquote></details>

Added special case where we send batch if all 3 are zeroed (so no need for `batchDefault`)


<!-- Sent from Reviewable.io -->

spiffyyeng commented on pkg/ccl/changefeedccl/sink_webhook.go:
```diff
@@ -81,15 +115,76 @@ type webhookSink struct {
+	batchCfg    batchConfig
+	// the i-th batchTracker should track and buffer events received by the
+	// i-th eventsChan. The guarantee here is per-key batching, so messages
+	// with the same key will be considered for the same batch.
```
plans to change this to make it global with 1 buffer: https://github.com/cockroachdb/cockroach/issues/68656

miretskiy reviewed (changes requested):
I think the above comment explaining the semantics of flushing behavior should be moved to commit messages.
Along with `CREATE CHANGEFEED .... WITH webhook_sink_config` examples included.

Also, the "per worker" bit should be included in PR description as well.

nit: Rephrase release note:
Release note (enterprise change): Add a `webhook_sink_config` JSON option to
configure batching and  flushing behavior, along with retry policy for webhook sink changefeed messages

> Reviewed 3 of 6 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68633#-:-Mgl8FxsAmyhAeaHx62Q:b-g7ni87)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @spiffyyeng and @stevendanna)

___
*[pkg/ccl/changefeedccl/sink_kafka.go, line 421 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/68633#-Mgl8mNfEdTbNg_hfxmN:-Mgl8mNfEdTbNg_hfxmO:bupm2oh) ([raw file](https://github.com/cockroachdb/cockroach/blob/0be267d177b4a7d000cc5b25baba77bc983bb015/pkg/ccl/changefeedccl/sink_kafka.go#L421)):*
> ```Go
> // JSONDuration is an aliased type for unmarshalling a Duration object
> // from JSON
> type JSONDuration time.Duration
> ```

I don't know if this needs to be exported; it is used only inside this package.
If you don't feel comfortable accessing JSON duration from another file in the same package, then maybe a better option would be to move this to e.g. cdcutils package (but that can be done later)
___
*[pkg/ccl/changefeedccl/sink_webhook.go, line 57 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68633#-Mgk6KEJDfMFIBJNgOU4:-Mgl80KT5dOxq5SIlP7G:b-7fhemp) ([raw file](https://github.com/cockroachdb/cockroach/blob/770aace7865c5ded1d3dd13d45b7a058ce39f2c1/pkg/ccl/changefeedccl/sink_webhook.go#L57)):*
<details><summary><i>Previously, spiffyyeng (Ryan Min) wrote…</i></summary><blockquote>

Done. (called size)
</blockquote></details>

I usually associate size with byte size
___
*[pkg/ccl/changefeedccl/sink_webhook.go, line 151 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/68633#-MglAReSERNbIxlBWqA3:-MglAReSERNbIxlBWqA4:bucf5bn) ([raw file](https://github.com/cockroachdb/cockroach/blob/0be267d177b4a7d000cc5b25baba77bc983bb015/pkg/ccl/changefeedccl/sink_webhook.go#L151)):*
<details><summary><i alt="summary">Quoted 21 lines of code…</i></summary>

> ```
> 	// wrapper structs to unmarshal json
> 	type retryConfig struct {
> 		Max     int          `json:",omitempty"`
> 		Backoff JSONDuration `json:",omitempty"`
> 	}
> 	// proper JSON schema for webhook sink config:
> 	// {
> 	//   "Flush": {
> 	//	   "Messages":  ...,
> 	//	   "Bytes":     ...,
> 	//	   "Frequency": ...,
> 	//   },
> 	//	 "Retry": {
> 	//	   "Max":     ...,
> 	//	   "Backoff": ...,
> 	//   }
> 	// }
> 	type webhookSinkConfig struct {
> 		Flush batchConfig `json:",omitempty"`
> 		Retry retryConfig `json:",omitempty"`
> 	}
> ```
</details>

Mind moving those structs outside of the function?
 I think they are reasonably important.
Also, not sure if you need to have separate batchConfig and retryConfig...
Try the following (but see if it works better/feels better):
```
type webhookSinkConfig struct 
    Flush  struct {
       ....
    }
    Retry struct {.  // JSON encoder automatically picks up Retry as the name, so -- all good.
       ...
    }
}
```
___
*[pkg/ccl/changefeedccl/sink_webhook.go, line 161 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/68633#-MglCWNS-CfSiqpr9mic:-MglCWNS-CfSiqpr9mid:b-vxr4r2) ([raw file](https://github.com/cockroachdb/cockroach/blob/0be267d177b4a7d000cc5b25baba77bc983bb015/pkg/ccl/changefeedccl/sink_webhook.go#L161)):*
<details><summary><i alt="summary">Quoted 7 lines of code…</i></summary>

> ```
> 		cfg = webhookSinkConfig{
> 			Flush: batchConfig{},
> 			Retry: retryConfig{
> 				Max:     retryCfg.MaxRetries,
> 				Backoff: JSONDuration(retryCfg.InitialBackoff),
> 			},
> 		}
> ```
</details>

do you need to initialize it like this?
Can you just start of with empty cfg, parse it and set whatever missing bits you need?
___
*[pkg/ccl/changefeedccl/sink_webhook.go, line 166 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/68633#-MglCI5k7mYAZ-xEjNw3:-MglCI5k7mYAZ-xEjNw4:b8pvjec) ([raw file](https://github.com/cockroachdb/cockroach/blob/0be267d177b4a7d000cc5b25baba77bc983bb015/pkg/ccl/changefeedccl/sink_webhook.go#L166)):*
<details><summary><i alt="summary">Quoted 5 lines of code…</i></summary>

> ```
> 		err = json.Unmarshal([]byte(configStr), &cfg)
> 		if err != nil {
> 			err = errors.Wrapf(err, "error unmarshalling json")
> 			return
> 		}
> ```
</details>

perhaps 
```
if err := json.Unmarshal(...); err != nil {
    return errors.Wrapf(....)
```
___
*[pkg/ccl/changefeedccl/sink_webhook.go, line 172 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/68633#-MglCAaU51qlT7cIJr3S:-MglCAaU51qlT7cIJr3T:b-gazam7) ([raw file](https://github.com/cockroachdb/cockroach/blob/0be267d177b4a7d000cc5b25baba77bc983bb015/pkg/ccl/changefeedccl/sink_webhook.go#L172)):*
> ```Go
> 	if cfg.Flush.Messages < 0 || cfg.Flush.Bytes < 0 || cfg.Flush.Frequency < 0 ||
> 		cfg.Retry.Max < 0 || cfg.Retry.Backoff < 0 {
> 		err = errors.Errorf("invalid option value %s, all config values must be non-negative", changefeedbase.OptWebhookSinkConfig)
> ```

I know you have "err" as the named parameter -- but you can still just
`return errors.Errorf`
___
*[pkg/ccl/changefeedccl/sink_webhook.go, line 178 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/68633#-MglCnmrBlK5Kk2-SdGd:-MglCnmrBlK5Kk2-SdGe:b-c5006s) ([raw file](https://github.com/cockroachdb/cockroach/blob/0be267d177b4a7d000cc5b25baba77bc983bb015/pkg/ccl/changefeedccl/sink_webhook.go#L178)):*
> ```Go
> 	// warning message if other batch values are set, but frequency is not
> 	if (cfg.Flush.Messages > 0 || cfg.Flush.Bytes > 0) && cfg.Flush.Frequency == 0 {
> 		log.Warningf(ctx, "flush frequency is not set, messages may never be sent")
> ```

This should be an error;

Or set default to something reasonable, but produce a warning.
___
*[pkg/ccl/changefeedccl/sink_webhook.go, line 362 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/68633#-MglE7va0ehDbq756yYT:-MglE7va0ehDbq756yYU:b-mrlroj) ([raw file](https://github.com/cockroachdb/cockroach/blob/0be267d177b4a7d000cc5b25baba77bc983bb015/pkg/ccl/changefeedccl/sink_webhook.go#L362)):*
> ```Go
> 	// frequency threshold has been reached
> 	case s.batchCfg.Frequency > 0 && timeutil.Since(f.bufferInitTime) >= time.Duration(s.batchCfg.Frequency):
> 		return true
> ```

I'm a bit confused by this choice.  It appears that frequency is honored only if the messages keep arriving?
And if you don't run w/ e.g. resolved option, then the last message will sit in the buffer for a potentially long time (until flush)... Not sure that's what flush frequency should do.

I think we need to have some sort of a go routine that basically runs on a timer; every time message arrives, it resets the timer so that messages will not be buffered for too long.
I imagine that we can add a "messageBatcher" function to the worker group (just like we add workerLoops).
EmitRow, instead of emitting directly to workers events channells, emits to "batcherChan"; batcher, of course
collects all messages into a single batch and decides if it's time to flush...
Pseudo code example below...

```
func (s *webhookSink) batchMessages() {
   var curentBatch  batchTracker // I would rename batchTracker to just batch...

   var ticker := timeutil.NewTimer()
  defer ticker.Stop()
   var flushCh <-chan time.Time. // nil initially

   select {
    case <-s.workerCtx.Done():
       return
    case msg := <- batchMessage:
       append_message_to_current_batch
       if TimeToSendBatchBecauseOfSize {
           SplitAndSendBatch(). // This function splits the batch into per worker batch message and sends that to appropriate worker.
            // If we flushed -- stop the timer.
            ticker.Stop()
            flushCh = nil. // wait until next message
      } else {
        // We didn't flush batch; so start the timer if needed 
        ticker.Reset(frequency)
        flushCh = ticker.C
    case <-flushCh: 
       // Time to send batch.  Note: reading from nil channel blocks forever -- that's good; 
       // We want to block until the first message arrives
       ticker.Read = true
       SplitAndSendBatch().
      //  Stop timer until next message arrives
      ticker.Stoop()
      flushCh = nil
}
```


<!-- Sent from Reviewable.io -->

spiffyy99 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68633)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @miretskiy, @spiffyyeng, and @stevendanna)

___
*[pkg/ccl/changefeedccl/sink_webhook.go, line 161 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/68633#-MglCWNS-CfSiqpr9mic:-Mgw7JKm2Hb20nKVOOaq:bpn9eln) ([raw file](https://github.com/cockroachdb/cockroach/blob/0be267d177b4a7d000cc5b25baba77bc983bb015/pkg/ccl/changefeedccl/sink_webhook.go#L161)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

> ```
> 		cfg = webhookSinkConfig{
> 			Flush: batchConfig{},
> 			Retry: retryConfig{
> 				Max:     retryCfg.MaxRetries,
> 				Backoff: JSONDuration(retryCfg.InitialBackoff),
> 			},
> 		}
> ```
do you need to initialize it like this?
Can you just start of with empty cfg, parse it and set whatever missing bits you need?
</blockquote></details>

Simplified a bit
___
*[pkg/ccl/changefeedccl/sink_webhook.go, line 166 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/68633#-MglCI5k7mYAZ-xEjNw3:-Mgw7NLb8NMTMwZqOHSO:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0be267d177b4a7d000cc5b25baba77bc983bb015/pkg/ccl/changefeedccl/sink_webhook.go#L166)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

> ```
> 		err = json.Unmarshal([]byte(configStr), &cfg)
> 		if err != nil {
> 			err = errors.Wrapf(err, "error unmarshalling json")
> 			return
> 		}
> ```
perhaps 
```
if err := json.Unmarshal(...); err != nil {
    return errors.Wrapf(....)
```

</blockquote></details>

Done.
___
*[pkg/ccl/changefeedccl/sink_webhook.go, line 172 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/68633#-MglCAaU51qlT7cIJr3S:-Mgw7O00E8qwPL-rIYJS:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0be267d177b4a7d000cc5b25baba77bc983bb015/pkg/ccl/changefeedccl/sink_webhook.go#L172)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

I know you have "err" as the named parameter -- but you can still just
`return errors.Errorf`
</blockquote></details>

Done.
___
*[pkg/ccl/changefeedccl/sink_webhook.go, line 178 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/68633#-MglCnmrBlK5Kk2-SdGd:-Mgw7OYG3QYgfdAlfcRv:bb74tv8) ([raw file](https://github.com/cockroachdb/cockroach/blob/0be267d177b4a7d000cc5b25baba77bc983bb015/pkg/ccl/changefeedccl/sink_webhook.go#L178)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

This should be an error;

Or set default to something reasonable, but produce a warning.
</blockquote></details>

Changed to error
___
*[pkg/ccl/changefeedccl/sink_webhook.go, line 362 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/68633#-MglE7va0ehDbq756yYT:-Mgw7QmI9f-GfXDpTtEj:bdynx87) ([raw file](https://github.com/cockroachdb/cockroach/blob/0be267d177b4a7d000cc5b25baba77bc983bb015/pkg/ccl/changefeedccl/sink_webhook.go#L362)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

I'm a bit confused by this choice.  It appears that frequency is honored only if the messages keep arriving?
And if you don't run w/ e.g. resolved option, then the last message will sit in the buffer for a potentially long time (until flush)... Not sure that's what flush frequency should do.

I think we need to have some sort of a go routine that basically runs on a timer; every time message arrives, it resets the timer so that messages will not be buffered for too long.
I imagine that we can add a "messageBatcher" function to the worker group (just like we add workerLoops).
EmitRow, instead of emitting directly to workers events channells, emits to "batcherChan"; batcher, of course
collects all messages into a single batch and decides if it's time to flush...
Pseudo code example below...

```
func (s *webhookSink) batchMessages() {
   var curentBatch  batchTracker // I would rename batchTracker to just batch...

   var ticker := timeutil.NewTimer()
  defer ticker.Stop()
   var flushCh <-chan time.Time. // nil initially

   select {
    case <-s.workerCtx.Done():
       return
    case msg := <- batchMessage:
       append_message_to_current_batch
       if TimeToSendBatchBecauseOfSize {
           SplitAndSendBatch(). // This function splits the batch into per worker batch message and sends that to appropriate worker.
            // If we flushed -- stop the timer.
            ticker.Stop()
            flushCh = nil. // wait until next message
      } else {
        // We didn't flush batch; so start the timer if needed 
        ticker.Reset(frequency)
        flushCh = ticker.C
    case <-flushCh: 
       // Time to send batch.  Note: reading from nil channel blocks forever -- that's good; 
       // We want to block until the first message arrives
       ticker.Read = true
       SplitAndSendBatch().
      //  Stop timer until next message arrives
      ticker.Stoop()
      flushCh = nil
}
```
</blockquote></details>

Changed the implementation to use a "batch worker" in between... might need some more testing though wrt/the timer.
___
*[pkg/ccl/changefeedccl/sink_kafka.go, line 421 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/68633#-Mgl8mNfEdTbNg_hfxmN:-Mgw7XC17za5pdB2AjIY:b-wpw4l8) ([raw file](https://github.com/cockroachdb/cockroach/blob/0be267d177b4a7d000cc5b25baba77bc983bb015/pkg/ccl/changefeedccl/sink_kafka.go#L421)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

I don't know if this needs to be exported; it is used only inside this package.
If you don't feel comfortable accessing JSON duration from another file in the same package, then maybe a better option would be to move this to e.g. cdcutils package (but that can be done later)
</blockquote></details>

Ah, my bad. forgot it was in the same package


<!-- Sent from Reviewable.io -->

stevendanna reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68633)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @miretskiy, @spiffyyeng, and @stevendanna)

___
*[pkg/ccl/changefeedccl/sink_webhook.go, line 404 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/68633#-MhDfxClDEEpSpGCKy_k:-MhDfxCmDzjUKuLyb13q:bugx1i9) ([raw file](https://github.com/cockroachdb/cockroach/blob/59354e80ae9aa6fadea2312f77cdd67e1d63a873/pkg/ccl/changefeedccl/sink_webhook.go#L404)):*
> ```Go
> 			}
> 		// timerCh will be nil if first message hasn't appeared, so won't trigger
> 		case <-batchTimer.C:
> ```

I'm not sure you are using this correctly.  I think that you'll want to make sure you set `batchTimer.Read = true` and then call `batchTimer.Reset()` at some point.   I'd read through the comments in the timerutil package, they have some useful examples.


<!-- Sent from Reviewable.io -->

spiffyy99 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68633)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @miretskiy, @spiffyyeng, and @stevendanna)

___
*[pkg/ccl/changefeedccl/sink_webhook.go, line 404 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/68633#-MhDfxClDEEpSpGCKy_k:-MhGkkZu9rEaS5t8MLkH:bdbxisd) ([raw file](https://github.com/cockroachdb/cockroach/blob/59354e80ae9aa6fadea2312f77cdd67e1d63a873/pkg/ccl/changefeedccl/sink_webhook.go#L404)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

I'm not sure you are using this correctly.  I think that you'll want to make sure you set `batchTimer.Read = true` and then call `batchTimer.Reset()` at some point.   I'd read through the comments in the timerutil package, they have some useful examples.
</blockquote></details>

added (note, we're using `timeutil.TimeSource` now so we can inject custom time logic for testing)


<!-- Sent from Reviewable.io -->

miretskiy reviewed (changes requested):
This looks really good!  Just have few minor nits/suggestions.

> Reviewed 1 of 6 files at r4, 1 of 3 files at r5, 5 of 5 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68633#-:-MhJSKk409MCKpkaeA9_:brcauit)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @miretskiy, @spiffyyeng, and @stevendanna)

___
*[-- commits, line 13 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/68633#-MhJS94W7vM3_tGlGnDA:-MhJS94W7vM3_tGlGnDB:b-eiguf7) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d0c4a47ed18e1beb5f96b3e19b01a2127f4a214/-- commits#L13)):*
I think you need to expand PR description to include some of the examples
of webhook_sink_config JSON...
___
*[pkg/ccl/changefeedccl/sink_webhook.go, line 411 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/68633#-MgxMbInBMm0b8xKUrM7:-MgxMbInBMm0b8xKUrM8:bdfs4kf) ([raw file](https://github.com/cockroachdb/cockroach/blob/029913b9470be3929534a98e29a8f3c06591cbf8/pkg/ccl/changefeedccl/sink_webhook.go#L411)):*
> ```Go
> 			batch := s.batchTracker.emptyBufferAndReset()
> 			ticker.Stop()
> 			if len(batch) > 0 {
> ```

sounds like that should always be true?
___
*[pkg/ccl/changefeedccl/sink_webhook.go, line 383 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/68633#-MhJKPHO--vCldu2ryAv:-MhJKPHO--vCldu2ryAw:b6mmpvv) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d0c4a47ed18e1beb5f96b3e19b01a2127f4a214/pkg/ccl/changefeedccl/sink_webhook.go#L383)):*
> ```Go
> func (s *webhookSink) batchWorker() {
> 	// batch messages are aggregated here and split into different workers
> 	// to be sent separately
> ```

nit: maybe move and rephrase this comment to be a function level comment:
```
// batchWorker aggregates messages into batches, and sends those batches to the workers
```
___
*[pkg/ccl/changefeedccl/sink_webhook_test.go, line 438 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/68633#-MhJRd9D1LF-wPMoevtd:-MhJRd9D1LF-wPMoevte:b3yawiw) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d0c4a47ed18e1beb5f96b3e19b01a2127f4a214/pkg/ccl/changefeedccl/sink_webhook_test.go#L438)):*
> ```Go
> 		// an empty batch to be sent
> 		mt.Advance(time.Hour)
> 		require.Equal(t, sinkDest.Latest(), "")
> ```

I'm not 100% sure on this, but it feels like checking for Latest() (and the fact that nothing is emitted) is a bit
racy -- i.e. do you need to wait for mt.Timers to be non-nil and the timer set to wait for 1 hour (like you do below)?
___
*[pkg/ccl/changefeedccl/sink_webhook_test.go, line 560 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/68633#-MhJRyJQ6HjKj4taXpkA:-MhJRyJQ6HjKj4taXpkB:b4mdufj) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d0c4a47ed18e1beb5f96b3e19b01a2127f4a214/pkg/ccl/changefeedccl/sink_webhook_test.go#L560)):*
> ```Go
> 
> 	// run tests with parallelism from 1-16 (1,2,4,8,16)
> 	for i := 1; i <= 16; i *= 2 {
> ```

perhaps that's too much? maybe just 1-4 or 5 would be okay.
Also, have you figured out why tests were taking long time?


<!-- Sent from Reviewable.io -->

spiffyy99 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68633)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @miretskiy, @spiffyyeng, and @stevendanna)

___
*[pkg/ccl/changefeedccl/sink_webhook.go, line 57 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68633#-Mgk6KEJDfMFIBJNgOU4:-MhJX8_q6QWGZHi1xXUG:bzs5sq) ([raw file](https://github.com/cockroachdb/cockroach/blob/770aace7865c5ded1d3dd13d45b7a058ce39f2c1/pkg/ccl/changefeedccl/sink_webhook.go#L57)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

I usually associate size with byte size
</blockquote></details>

changed to length
___
*[pkg/ccl/changefeedccl/sink_webhook.go, line 411 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/68633#-MgxMbInBMm0b8xKUrM7:-MhJY7uf3EQORB4NLXiR:bot6tbw) ([raw file](https://github.com/cockroachdb/cockroach/blob/029913b9470be3929534a98e29a8f3c06591cbf8/pkg/ccl/changefeedccl/sink_webhook.go#L411)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

sounds like that should always be true?
</blockquote></details>

we discussed this before, it won't be true always. after sending a batch, we don't reset immediately. so on the next batch (which is still empty), the channel will fire, so we need to have this check to make sure an empty batch doesn't send.


<!-- Sent from Reviewable.io -->

spiffyy99 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68633)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @miretskiy, @spiffyyeng, and @stevendanna)

___
*[pkg/ccl/changefeedccl/sink_webhook.go, line 411 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/68633#-MgxMbInBMm0b8xKUrM7:-MhJYNP81hTO4_iYwZek:b-wx6jj3) ([raw file](https://github.com/cockroachdb/cockroach/blob/029913b9470be3929534a98e29a8f3c06591cbf8/pkg/ccl/changefeedccl/sink_webhook.go#L411)):*
<details><summary><i>Previously, spiffyyeng (Ryan Min) wrote…</i></summary><blockquote>

we discussed this before, it won't be true always. after sending a batch, we don't reset immediately. so on the next batch (which is still empty), the channel will fire, so we need to have this check to make sure an empty batch doesn't send.
</blockquote></details>

* could fire, won't fire for sure necessarily


<!-- Sent from Reviewable.io -->

spiffyy99 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68633)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @miretskiy, @spiffyyeng, and @stevendanna)

___
*[pkg/ccl/changefeedccl/sink_webhook_test.go, line 438 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/68633#-MhJRd9D1LF-wPMoevtd:-MhJp-nB9PSsCrdY-KQf:bd30ifx) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d0c4a47ed18e1beb5f96b3e19b01a2127f4a214/pkg/ccl/changefeedccl/sink_webhook_test.go#L438)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

I'm not 100% sure on this, but it feels like checking for Latest() (and the fact that nothing is emitted) is a bit
racy -- i.e. do you need to wait for mt.Timers to be non-nil and the timer set to wait for 1 hour (like you do below)?
</blockquote></details>

good catch
___
*[pkg/ccl/changefeedccl/sink_webhook_test.go, line 560 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/68633#-MhJRyJQ6HjKj4taXpkA:-MhJiU1FBa8eKLQblwGR:b-o9azjy) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d0c4a47ed18e1beb5f96b3e19b01a2127f4a214/pkg/ccl/changefeedccl/sink_webhook_test.go#L560)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

perhaps that's too much? maybe just 1-4 or 5 would be okay.
Also, have you figured out why tests were taking long time?
</blockquote></details>

will update to be 1-4 (that should reduce the time a bit, considering all the workers that need to be created now). don't think the time is too out of the ordinary though considering we have 5 parallelisms here and 6 subtests for each (30 in total), creating those goroutines for each sink invocation might take some time 

if it helps, from my observations, doesn't look like we're waiting for the timer for too long


<!-- Sent from Reviewable.io -->

spiffyy99 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68633)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @miretskiy, @spiffyyeng, and @stevendanna)

___
*[-- commits, line 13 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/68633#-MhJS94W7vM3_tGlGnDA:-MhJq2a46KDX5W4p8br9:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d0c4a47ed18e1beb5f96b3e19b01a2127f4a214/-- commits#L13)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

I think you need to expand PR description to include some of the examples
of webhook_sink_config JSON...
</blockquote></details>

Done.
___
*[pkg/ccl/changefeedccl/sink_webhook.go, line 151 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/68633#-MglAReSERNbIxlBWqA3:-MhJq9IRDjPwJom5UIIO:b-84chmc) ([raw file](https://github.com/cockroachdb/cockroach/blob/0be267d177b4a7d000cc5b25baba77bc983bb015/pkg/ccl/changefeedccl/sink_webhook.go#L151)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

> ```
> 	// wrapper structs to unmarshal json
> 	type retryConfig struct {
> 		Max     int          `json:",omitempty"`
> 		Backoff JSONDuration `json:",omitempty"`
> 	}
> 	// proper JSON schema for webhook sink config:
> 	// {
> 	//   "Flush": {
> 	//	   "Messages":  ...,
> 	//	   "Bytes":     ...,
> 	//	   "Frequency": ...,
> 	//   },
> 	//	 "Retry": {
> 	//	   "Max":     ...,
> 	//	   "Backoff": ...,
> 	//   }
> 	// }
> 	type webhookSinkConfig struct {
> 		Flush batchConfig `json:",omitempty"`
> 		Retry retryConfig `json:",omitempty"`
> 	}
> ```
Mind moving those structs outside of the function?
 I think they are reasonably important.
Also, not sure if you need to have separate batchConfig and retryConfig...
Try the following (but see if it works better/feels better):
```
type webhookSinkConfig struct 
    Flush  struct {
       ....
    }
    Retry struct {.  // JSON encoder automatically picks up Retry as the name, so -- all good.
       ...
    }
}
```
</blockquote></details>

problem is, we can't exactly use the same config struct to unmarshal JSON and pass to the sink struct (we have to use retry.Options for the sink which doesn't fit the JSON schema). batchConfig is simply a way to reuse the schema between both.
___
*[pkg/ccl/changefeedccl/sink_webhook.go, line 383 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/68633#-MhJKPHO--vCldu2ryAv:-MhJqovw1SgINEbnrKnd:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d0c4a47ed18e1beb5f96b3e19b01a2127f4a214/pkg/ccl/changefeedccl/sink_webhook.go#L383)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

nit: maybe move and rephrase this comment to be a function level comment:
```
// batchWorker aggregates messages into batches, and sends those batches to the workers
```
</blockquote></details>

done


<!-- Sent from Reviewable.io -->

miretskiy reviewed (commented):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68633#-:-MhKZ9IW4obgyodaAXwq:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @miretskiy, @spiffyyeng, and @stevendanna)



<!-- Sent from Reviewable.io -->

spiffyyeng commented:
TFTR!

spiffyyeng commented:
bors r+

spiffyy99 committed:
changefeedccl: webhook_sink_config option to configure batching and retries

Previously, retry behavior was preset and batching was not supported. This
change adds configuration JSON to change retry behavior and batch sink
messages by count, bytes, and timeouts.

Batching is done per-worker (and thus per-key), so each message with the
same key will be considered for the same batch.
There are two config params for retries (Max and Backoff), which are 3
and 500ms by default if not included.

There are three config params for batching (Messages, Bytes and
Frequency), all of which are assumed to be zero if not included.
{
   "Flush": {
     "Messages":  ...,
     "Bytes":     ...,
     "Frequency": ...,
   },
   "Retry": {
     "Max":     ...,
     "Backoff": ...,
  }
}

For batching params specifically, there are a few caveats:

Example:  If all three are zeroed, the sink will send batches on
every message. (this is the default behavior)
{
   "Flush": {
     "Messages":  0,
     "Bytes":     0,
     "Frequency": "0s"
   }
}

Example: If at least one is nonzero, the other zero/omitted values will be
considered infinity (the sink requires setting a frequency in this case).
{
   "Flush": {
     "Messages":  100,
     "Frequency": "30s"
   }
}

Resolves #67804

Release note (enterprise change): Add a webhook_sink_config JSON option to
configure batching and flushing behavior, along with retry behavior for
webhook sink changefeed messages.

spiffyyeng commented:
bors retry
